qiskit/quantum_info/operators/symplectic/base_pauli.py

```diff
@@ -0,0 +1,639 @@
+        """
+        self._z = z
+        self._x = x
+        self._phase = phase
+        self._num_paulis, num_qubits = self._z.shape
+        super().__init__(num_qubits=num_qubits)
+
+    def copy(self):
+        """Make a deep copy of current operator."""
+        # Deepcopy has terrible performance on objects with Numpy arrays
+        # attributes so we make a shallow copy and then manually copy the
+        # Numpy arrays to efficiently mimic a deepcopy
+        ret = copy.copy(self)
+        ret._z = self._z.copy()
+        ret._x = self._x.copy()
+        ret._phase = self._phase.copy()
+        return ret
+
+    # ---------------------------------------------------------------------
+    # BaseOperator methods
+    # ---------------------------------------------------------------------
+
+    def tensor(self, other):
+        """Return the tensor product Pauli self ⊗ other.
+
+        Args:
+            other (BasePauli): another Pauli.
+
+        Returns:
+            BasePauli: the tensor product Pauli.
+        """
+        z = np.hstack([self._stack(other._z, self._num_paulis), self._z])
+        x = np.hstack([self._stack(other._x, self._num_paulis), self._x])
+        phase = np.mod(self._phase + other._phase, 4)
+        return BasePauli(z, x, phase)
+
+    def expand(self, other):
+        """Return the tensor product Pauli other ⊗ self.
+
+        Args:
+            other (BasePauli): another Pauli.
+
+        Returns:
+            BasePauli: the tensor product Pauli.
+        """
+        z = np.hstack([self._z, self._stack(other._z, self._num_paulis)])
+        x = np.hstack([self._x, self._stack(other._x, self._num_paulis)])
+        phase = np.mod(self._phase + other._phase, 4)
+        return BasePauli(z, x, phase)
+
+    # pylint: disable=arguments-differ
+    def compose(self, other, qargs=None, front=True, inplace=False):
+        """Return the composition of Paulis self∘other.
+
+        Note that this discards phases.
+
+        Args:
+            other (BasePauli): another Pauli.
+            qargs (None or list): qubits to apply dot product on (default: None).
+            front (bool): If True use `dot` composition method (default: False).
+            inplace (bool): If True update in-place (default: False).
+
+        Returns:
+            BasePauli: the output Pauli.
+
+        Raises:
+            QiskitError: if number of qubits of other does not match qargs.
+        """
+        # Validation
+        if qargs is None and other.num_qubits != self.num_qubits:
+            raise QiskitError(
+                "other {} must be on the same number of qubits.".format(
+                    type(self).__name__))
+
+        if qargs and other.num_qubits != len(qargs):
+            raise QiskitError(
+                "Number of qubits of the other {} does not match qargs.".format(
+                    type(self).__name__))
+
+        if len(other) not in [1, len(self)]:
```

Comment:
`BasePauli` does not have `__len__`. It is implemented in `Pauli` instead. How about adding the following?
```python
def __len__(self):
    return self._num_qubits
```

Reply:
I would probably say len of `BasePauli` is the number of `Pauli`s, not the number of qubits of each Pauli, since this is the leading dimension in the internal arrays.

Reply:
I see. The following [error message](https://github.com/chriseclectic/qiskit-terra/blob/qi/pauliop/qiskit/quantum_info/operators/symplectic/base_pauli.py#L122) says as follows.
```
            raise QiskitError("Incompatible BasePaulis. Second list must "
                              "either have 1 or the same number of Paulis.")
```
`len` should be number of Paulis.

But, `Pauli` has [`__len__` as follows](https://github.com/chriseclectic/qiskit-terra/blob/qi/pauliop/qiskit/quantum_info/operators/symplectic/pauli.py#L261
). Isn't it confusing?
```python
    def __len__(self):
        """Return the number of qubits in the Pauli."""
        return self.num_qubits
```

Reply:
In principle that error message should never get seen when used through `Pauli` class (it was intended for PauliList class). Either way I think you found a bug in the BasePauli class since `len` there is not defined as you say, and it's just lucky this worked for the regular single `Pauli`!